### ISSUE TITLE
Pipeline mode has Thread-safety bug

### ISSUE THREAD

**Issue by zhaozhiguang** [spring-projects/spring-data-redis#2653]
#### Current Behavior

Use parallelStream for concurrent operations. The amount of data is about 300k. After execution, all results will be added to the pipe. However, this pipe is of the ArrayList type when initializing, so some results are not added to the pipe. Then, NullPointerException will be thrown when traversing the pipe when closePipeline()

<details>
<summary>Stack trace</summary>

```java
java.lang.NullPointerException:null
 at org.springframework.data.redis.connection.lettuce.LettuceConnection.closePipeline(LettuceConnection.java:515)
...
```

</details>

#### Input Code

<details>
<summary>Input Code</summary>

```java
redisTemplate.executePipeLined(connection -> {
	objectives.parallelStream().forEach(obj -> {
		connection.setCommands().sAdd(rawKey,rawValue);
	});
})
```

</details>

#### Environment

- Spring-Data-Redis version(s): [e.g. 2.5.6]
- Redis version: [e.g. 6.2.5]

**Comment by jxblum** [comment]
@zhaozhiguang -

Perhaps you can explain what `objectives` is from your "_Input Code_":

```java
redisTemplate.executePipelined(connection -> {
	objectives.parallelStream().forEach(obj -> {
		connection.setCommands().sAdd(rawKey, rawValue);
	});
})
```

For my explanation, I assume (and I am reasonably certain) that:

1) `objectives` is some type of `Collection`
2) And, you are calling `parallelStream()` ([Javadoc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collection.html#parallelStream())) on this `Collection`
3) And, the `Collection` implementation supports parallel processing of the `Stream` returned by the `Collection` that is also the source for the elements contained in the `Stream`.

First of all, the Spring Data Redis `RedisConnection` interface is **not** documented as being Thread-safe, not in the [reference documentation](https://docs.spring.io/spring-data/redis/docs/current/reference/html/#redis:connectors) and not in [Javadoc](https://docs.spring.io/spring-data/redis/docs/current/api/org/springframework/data/redis/connection/RedisConnection.html).

Thread-safety is certainly **not** the case for _Jedis_. Under-the-hood, a Spring Data Redis `JedisConnection` ([Javadoc](https://docs.spring.io/spring-data/redis/docs/current/api/org/springframework/data/redis/connection/jedis/JedisConnection.html); [source](https://github.com/spring-projects/spring-data-redis/blob/2.7.14/src/main/java/org/springframework/data/redis/connection/jedis/JedisConnection.java#L71)) uses a _Jedis_ client driver, `Jedis` instance, which is **not** Thread-safe (see [here](https://github.com/redis/jedis/wiki/Getting-started#using-jedis-in-a-multithreaded-environment)).

For _Lettuce_, while the underlying, native connection is Thread-safe, as documented in Spring Data Redis's [reference documentation](https://docs.spring.io/spring-data/redis/docs/current/reference/html/#redis:connectors:lettuce) as well as the _Lettuce_ client driver docs (see [RedisClient](https://lettuce.io/core/release/api/io/lettuce/core/RedisClient.html) and `StatefulConnection`, such as [StatefulRedisConnection](https://lettuce.io/core/release/api/io/lettuce/core/api/StatefulRedisConnection.html)), it is **not** necessarily `true` that a Spring Data Redis `LuttuceConnection` itself is Thread-safe.

> NOTE: The `List` (implementation) used in pipelining ([here](https://github.com/spring-projects/spring-data-redis/blob/2.7.14/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnection.java#L117), and specifically constructed, [here](https://github.com/spring-projects/spring-data-redis/blob/2.7.14/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnection.java#L511), when opening the pipeline) is not the only non-Thread-safe data structure used by `LettuceConnection`. `LettuceConnection` also uses a non-Thread-safe `Queue` (backed by a `LinkedList` implementation) for collecting TX results (see [here](https://github.com/spring-projects/spring-data-redis/blob/2.7.14/src/main/java/org/springframework/data/redis/connection/lettuce/LettuceConnection.java#L119)).

When a Java stream is parallelized (such as by calling `parallelStream()`), the stream is partitioned into sub-streams that are then processed in parallel (or concurrently) using multiple Threads created from the _ForkJoin_ framework (pool).  

Therefore, a `LettuceConnection` is being accessed from multiple Threads based on how you structured your code.

> TIP: It is important to understand the implications of "parallel" streams (and parallel processing in general). You can find documentation on this matter from [Oracle](https://docs.oracle.com/javase/tutorial/collections/streams/parallelism.html), as well as in this nice [blog post](https://www.baeldung.com/java-when-to-use-parallel-stream) from Baeldung.

As a temporary workaround, you could have simply, instead, sequentially processed the elements (`objectives`) in the `Collection` by doing:

```java
redisTemplate.executePipelined(connection -> {
	objectives.forEach(obj -> {
		connection.setCommands().sAdd(rawKey, rawValue);
	});
})
```

You are already benefiting from using "Redis pipelining", as [documented](https://redis.io/docs/manual/pipelining/).

Alternatively, although I am not entirely certain in this case, you may be able to restructure your code by doing:

```java
objectives.parallelStream().forEach(obj -> {
        redisTemplate.executePipelined(operations -> {
		operations.opsForSet().add(rawKey, rawValue);
	});
})

```

In this case, you are calling `RedisTemplate.executePipelined(:SessionCallback<?>)` ([Javadoc](https://docs.spring.io/spring-data/redis/docs/current/api/org/springframework/data/redis/core/RedisOperations.html#executePipelined(org.springframework.data.redis.core.SessionCallback))).

Also, see the [Javadoc](https://docs.spring.io/spring-data/redis/docs/current/api/org/springframework/data/redis/core/SessionCallback.html) for `SessionCallback`.

Of course, `SessionCallback` seems to be intended for use in transactions. Still, like transactions, pipelining works against a dedicated connection to send a batch of Redis commands in a single request (without the atomic guarantee as is for TX).

> NOTE: `RedisTemplate` is Thread-safe (see [Javadoc](url), and [ref doc](https://docs.spring.io/spring-data/redis/docs/current/reference/html/#redis:template)).

Let me know what you think.

In the meantime, I will speak to the team about our Thread-safety guarantees with regard to Spring Data Redis `RedisConnections`.

**Comment by jxblum** [comment]
@zhaozhiguang -

Also be aware that Spring Data Redis `2.5.x` (last version was `2.5.12`) is no longer in OSS support (see [here](https://spring.io/projects/spring-data-redis#support)). Even commercial support ends in a couple weeks, on 2023-August-14.

You should minimally upgrade to Spring Data Redis `2.7.x` (latest version is `2.7.14`). Although, be aware that SD Redis `2.7.x` OSS support ends in November, 2023, after which, your team will need to purchase a commercial support from VMware by purchasing a VMware Spring Runtime (VSR; see [here](https://tanzu.vmware.com/spring-runtime)) license.

I would encourage you to upgrade to Spring Data Redis `3.1.x` as your earliest convenience.  Spring Data Redis `3.x` does require a Java 17 baseline, uses Spring Framework 6, and is pulled in by Spring Boot 3.

**Comment by jxblum** [comment]
NOTE: I edited my original, first [comment](https://lettuce.io/core/release/api/io/lettuce/core/api/StatefulConnection.html) above.

**Comment by jxblum** [comment]
@zhaozhiguang  - Following up to my previous comments, I have additional feedback for you.


First, I wrote a few [integration tests](https://github.com/jxblum/redis-experiments/blob/main/spring-data-redis-experiments/src/test/java/io/vmware/spring/data/redis/client/tests/ConcurrentRedisPipeliningIntegrationTests.java#L131-L199) to reproduce and analyze the problem.  A few things should stand out here:

1) The `storeInRedisUsingParallelStreamInPipeline()` [test case](https://github.com/jxblum/redis-experiments/blob/main/spring-data-redis-experiments/src/test/java/io/vmware/spring/data/redis/client/tests/ConcurrentRedisPipeliningIntegrationTests.java#L131-L152) is precisely your scenario.

2) I could only reproduce the NPE after using a data set size of `500,000` elements ([here](https://github.com/jxblum/redis-experiments/blob/main/spring-data-redis-experiments/src/test/java/io/vmware/spring/data/redis/client/tests/ConcurrentRedisPipeliningIntegrationTests.java#L107)), running Oracle JDK `17.0.7` in macOS Ventura `13.5`  on _Intel_).

3) The test case both passed and failed (throwing the NPE, as [documented](https://github.com/jxblum/redis-experiments/blob/main/spring-data-redis-experiments/src/test/java/io/vmware/spring/data/redis/client/tests/ConcurrentRedisPipeliningIntegrationTests.java#L61-L78)) across different and several integration test runs. It does not consistently fail or pass, though it passed more often than it failed.

4) The **failure** was not unexpected (nor was a "**pass**" outcome) given the nature of "_race conditions_" when using non-Thread-safe components, such as Redis connections, and in particular, the Spring Data Redis `LettuceConnection` class.

5) I compared your arrangement to using a **sequential stream** for processing (storing) the elements in a Redis Set, in this [test case](https://github.com/jxblum/redis-experiments/blob/main/spring-data-redis-experiments/src/test/java/io/vmware/spring/data/redis/client/tests/ConcurrentRedisPipeliningIntegrationTests.java#L158-L172).

6) The **ordering** between the [parallel](https://github.com/jxblum/redis-experiments/blob/main/spring-data-redis-experiments/src/test/java/io/vmware/spring/data/redis/client/tests/ConcurrentRedisPipeliningIntegrationTests.java#L131-L134) and [sequential](https://github.com/jxblum/redis-experiments/blob/main/spring-data-redis-experiments/src/test/java/io/vmware/spring/data/redis/client/tests/ConcurrentRedisPipeliningIntegrationTests.java#L132-L134) test cases in my integration test class was deliberate.  Effectively, I idid NOT want the "sequential" test case to have the luxury of a warm JVM with optimizations when compared to the "parallel" test case, where the JVM might possibly recognize similar codepaths and the JIT kicking in.

7) Which, I then added [time measurements](https://github.com/jxblum/redis-experiments/blob/main/spring-data-redis-experiments/src/test/java/io/vmware/spring/data/redis/client/tests/ConcurrentRedisPipeliningIntegrationTests.java#L138) (and [timing for sequential execution](https://github.com/jxblum/redis-experiments/blob/main/spring-data-redis-experiments/src/test/java/io/vmware/spring/data/redis/client/tests/ConcurrentRedisPipeliningIntegrationTests.java#L138)) to compare the sequential and parallel stream use inside Redis Pipelined connections.

8) Finally, I [logged the Thread names in use](https://github.com/jxblum/redis-experiments/blob/main/spring-data-redis-experiments/src/test/java/io/vmware/spring/data/redis/client/tests/ConcurrentRedisPipeliningIntegrationTests.java#L149) when testing parallel stream execution inside Redis Pipelined connections to ensure the `Collection` supported parallel streams, and in fact, different Threads were used to process the `Stream` returned by the `Collection`. The output should look similar to:

```txt
2023-08-01T14:42:20.145-07:00 DEBUG 57021 --- [           main] .s.d.r.c.t.AbstractRedisIntegrationTests : PIPELINE-STREAM THREADS [[ForkJoinPool.commonPool-worker-1, ForkJoinPool.commonPool-worker-10, ForkJoinPool.commonPool-worker-11, ForkJoinPool.commonPool-worker-2, ForkJoinPool.commonPool-worker-3, ForkJoinPool.commonPool-worker-4, ForkJoinPool.commonPool-worker-5, ForkJoinPool.commonPool-worker-6, ForkJoinPool.commonPool-worker-7, ForkJoinPool.commonPool-worker-8, ForkJoinPool.commonPool-worker-9, main]]
```

You may be surprised to learn that the "**sequential**" execution outperformed the "**parallel**" execution in my testing. Of course, this is not unexpected. Parallelization (or premature optimization) does not always lead to performance improvements. It depends on the environment and measurements are always required to optimize correctly, if at all.

I'd say, in your case, with only a 300K element data set size, it won't make much of a difference to optimize through the use parallelization. It will also most definitely vary by environment. And, most of the time is likely to be eaten up by network activity, anyway, which can vary from moment to moment depending on the amount of network traffic.

In my test run, although local, the difference between sequential and parallel execution on 500,000 elements was not significant, with sequential execution running slightly faster even though it was run first (possibly void of JIT compilation (optimizations in general)):

Using Parallel Stream: **10,918 ms**
Using Sequential Stream: **10,269 ms**

Of course, JVM optimizations can vary between run invocations as well.

**Comment by jxblum** [comment]
Finally, I'd also add that you should NOT use the alternative approach I recommend, which I also captured in a [test case](https://github.com/jxblum/redis-experiments/blob/main/spring-data-redis-experiments/src/test/java/io/vmware/spring/data/redis/client/tests/ConcurrentRedisPipeliningIntegrationTests.jatva#L183-L200).

This performed very poorly, and when you trace down (into Spring Data Redis), you realize that you are not really using Redis Pipelined connections effectively since the `SessionCallback` is not a true session in the sense of coupling a connection to a `Thread` using `ThreadLocals` in the same manner that Thread-context transactions are managed.

**Comment by zhaozhiguang** [comment]
Yes, I have also tested both concurrent and serial modes. Serial performance will be better, because I read Lettuce's official website that Thread safety is, of course, Spring Data Redis is also Thread safety

**Comment by jxblum** [comment]
Not all classes and components in Spring Data Redis are Thread-safe, as I have discussed above.

For instance, clearly the `RedisConnection` classes are NOT Thread-safe. `JedisConnection` is definitely not Thread-safe since _Jedis_ itself is not Thread-safe ([this](https://github.com/redis/jedis/wiki/Getting-started#using-jedis-in-a-multithreaded-environment)).

While the Spring Data Redis `RedisTemplate` is [documented](https://docs.spring.io/spring-data/redis/docs/current/api/org/springframework/data/redis/core/RedisTemplate.html) to be Thread-safe, it depends on how it is used.

So long as the `RedisTemplate` is acquired from the Spring container (i.e. injected into your application components and beans managed by the Spring container), and NOT re-configured, it will remain Thread-safe. It can be safely shared across multiple Threads, even though `RedisTemplate` uses connections, because it acquires, uses and then releases connections per operation, thereby maintaining Thread-safety (why connection pooling helps here). If it kept a reference to, and simply reused a single connection, then `RedisTemplate` would no longer be Thread-safe. If the template configuration is changed after it is acquired from (or injected by) the Spring container, then it is no longer Thread-safe since it carries non-volatile references to other objects, like `RedisSerializers` used to de/serialize keys/values, the `ClassLoader` used on deserialization, and so on ([all this](https://github.com/spring-projects/spring-data-redis/blob/2.7.14/src/main/java/org/springframework/data/redis/core/RedisTemplate.java#L93-L106); mileage on operation interface delegates varies).

> NOTE: Non-volatile and non-atomic references to mutable state is not Thread-safe, and even if they were volatile/atomic, the objects to which they refer much also be Thread-safe (or effectively immutable) for the template to be Thread-safe.

In summary, the "Thread-safety" guarantee of `RedisTemplate` hinges on "initialization-safety" and nothing changing after initialization.

**Comment by jxblum** [comment]
Two things. In general...

1) Not every class that consists entirely of or uses Thread-safe objects is itself Thread-safe. For example, think compound actions on multiple, independent, Thread-safe objects.

2) And, not every class consisting of or using non-Thread-safe classes is itself, not Thread-safe. For example, think immutable, or effectively immutable, and safely-initialized components.

There are many factors for determining the Thread-safety guarantees of what constitutes a "Thread-safe" class.
